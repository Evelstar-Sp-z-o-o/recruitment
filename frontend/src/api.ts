import type { Post } from './types';

export const BASE_API_URL = 'http://localhost:3000/api';

const getPosts = async () => {
  const ENDPOINT = '/posts';

  const res = await fetch(BASE_API_URL + ENDPOINT);

  if (!res.ok) {
    throw new Error('Something went wrong when loading Posts list.');
  }

  const data = await res.json();

  return data as unknown as Post[];
};

const getSinglePost = async (postId: string) => {
  const ENDPOINT = `/posts/${postId}`;

  const res = await fetch(BASE_API_URL + ENDPOINT);

  if (!res.ok) {
    throw new Error('Something went wrong when loading Post details.');
  }

  const data = await res.json();
  return data as unknown as Post;
};

const deletePost = async (postId: string) => {
  const ENDPOINT = `/posts/${postId}`;

  const res = await fetch(BASE_API_URL + ENDPOINT, {
    method: 'DELETE',
  });

  if (!res.ok) {
    throw new Error('Something went wrong when deleting post.');
  }

  const data = await res.json();
  return data as unknown as Post;
};

const updatePost = async (postId: string, postData: Partial<Post>) => {
  const ENDPOINT = `/posts/${postId}`;

  const res = await fetch(BASE_API_URL + ENDPOINT, {
    method: 'PATCH',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(postData),
  });

  if (!res.ok) {
    throw new Error('Something went wrong when updating post.');
  }

  const data = await res.json();
  return data as unknown as Post;
};

const createPost = async (postData: Partial<Post>) => {
  const ENDPOINT = `/posts`;

  const res = await fetch(BASE_API_URL + ENDPOINT, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(postData),
  });

  if (!res.ok) {
    throw new Error('Something went wrong when creating post.');
  }

  const data = await res.json();

  return data as unknown as Post;
};

export { getPosts, getSinglePost, updatePost, createPost, deletePost };
