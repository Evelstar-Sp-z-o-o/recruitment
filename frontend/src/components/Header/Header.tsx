import FlutterDashIcon from '@mui/icons-material/FlutterDash';
import { Box, Typography } from '@mui/material';

const Header = () => {
  return (
    <Box
      sx={{
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
        margin: '10px auto 20px',
        gap: '15px',
        borderRadius: '20px',
        padding: '5px',
        backgroundColor: 'white',
        boxShadow: '8px 8px 24px 0px rgba(66, 68, 90, 1)',
      }}
    >
      <Typography sx={{ fontSize: '40px', color: '#2b7ad4' }}>QUACKER</Typography>
      <FlutterDashIcon
        sx={{
          fontSize: '40px',
        }}
        color="primary"
      />
    </Box>
  );
};

export default Header;
