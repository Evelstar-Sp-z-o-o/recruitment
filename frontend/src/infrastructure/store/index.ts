import { useDispatch, useSelector } from 'react-redux';

import { configureStore } from '@reduxjs/toolkit';

import apiService from './services/apiService';
import userSlice from './slices/user';

const store = configureStore({
  reducer: {
    [apiService.reducerPath]: apiService.reducer,
    user: userSlice.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(apiService.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector = useSelector.withTypes<RootState>();

export default store;
