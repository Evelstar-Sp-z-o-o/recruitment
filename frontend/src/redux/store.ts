import { configureStore } from '@reduxjs/toolkit';

import responseModalReducer from './responseModalSlice';

const store = configureStore({
  reducer: {
    responseModal: responseModalReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
