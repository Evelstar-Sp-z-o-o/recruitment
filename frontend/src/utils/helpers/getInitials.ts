export const getInitials = (user: string) => {
  const initials: string[] = [];
  if (user.length > 0) {
    const name = user.substring(0, user.indexOf('@'));
    for (let i = 0; i < name.split('.').length; i++) {
      initials.push(name.split('.')[i][0].toUpperCase());
    }
  }

  return initials.join('');
};
