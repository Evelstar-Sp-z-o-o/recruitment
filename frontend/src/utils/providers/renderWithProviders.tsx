import { JSX, PropsWithChildren } from 'react';

import { render } from '@testing-library/react';

import AppProviders from './AppProviders.tsx';

export const renderWithProviders = (children: PropsWithChildren<JSX.Element>) => {
  return render(<AppProviders>{children}</AppProviders>);
};
