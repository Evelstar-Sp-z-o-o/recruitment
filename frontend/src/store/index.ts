import { postsApi } from '@/src/store/api/posts';
import { configureStore, createSlice } from '@reduxjs/toolkit';

const initialUser: null | string = null;
const initialLogin = false;

const userSlice = createSlice({
  name: 'user',
  initialState: initialUser,
  reducers: {
    setUser(_state, action) {
      return action.payload;
    },
  },
});

const loginSlice = createSlice({
  name: 'login',
  initialState: initialLogin,
  reducers: {
    setLogin(_state, action) {
      return action.payload;
    },
  },
});

export const { setUser } = userSlice.actions;
export const { setLogin } = loginSlice.actions;

export * from './api/posts.ts';

export const store = configureStore({
  reducer: {
    [postsApi.reducerPath]: postsApi.reducer,
    user: userSlice.reducer,
    login: loginSlice.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(postsApi.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
