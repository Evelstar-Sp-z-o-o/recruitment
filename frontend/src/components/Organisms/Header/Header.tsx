import { FC } from 'react';
import { useSelector } from 'react-redux';
import { useLocation } from 'react-router-dom';

import background from '@/src/assets/bgr.webp';
import Logo from '@/src/assets/logo.svg?react';
import Unknown from '@/src/assets/noUserIcon.svg';
import Avt from '@/src/assets/userIcon.svg';
import { RootState } from '@/src/store';
import { getInitials } from '@/src/utils/helpers/getInitials';
import AppBar from '@mui/material/AppBar';
import Avatar from '@mui/material/Avatar';
import IconButton from '@mui/material/IconButton';
import SvgIcon from '@mui/material/SvgIcon';
import Typography from '@mui/material/Typography';
import { green } from '@mui/material/colors';

const Header: FC = ({ toggleMenu }: { toggleMenu: () => void }) => {
  const user: string = useSelector<RootState>((state) => state.user);
  const location = useLocation();
  const isProfile: boolean = location.pathname.slice(1) === 'profile';

  const StyledLogo = {
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)',
    fontSize: '3rem',
  } as const;

  const StyledAvatarButton = {
    width: isProfile ? '6rem' : '4rem',
    height: isProfile ? '6rem' : '4rem',
    position: isProfile ? 'absolute' : 'static',
    top: isProfile ? '100%' : 'unset',
    left: isProfile ? '50%' : 'unset',
    transform: isProfile ? 'translate(-50%, -50%)' : 'unset',
  } as const;

  const StyledAvatarIcon = {
    width: isProfile ? '4rem' : '2.5rem',
    height: isProfile ? '4rem' : '2.5rem',
    bgcolor: isProfile ? green[700] : 'transparent',
  } as const;

  const StyleHeaderWrapper = {
    p: '2rem',
    mb: isProfile ? '4rem' : '0',
    position: 'relative',
    minHeight: isProfile ? '10rem' : 'auto',
    background: isProfile ? `url(${background}) center / cover no-repeat` : 'unset',
  } as const;

  return (
    <AppBar position="static" color="default" sx={StyleHeaderWrapper}>
      <IconButton onClick={toggleMenu(true)} sx={StyledAvatarButton}>
        <Avatar alt="User" src={isProfile ? null : user ? Avt : (Unknown as string)} sx={StyledAvatarIcon}>
          {isProfile ? getInitials(user) : null}
        </Avatar>
      </IconButton>
      {isProfile ? (
        <Typography
          variant="h5"
          component="h2"
          sx={{ position: 'absolute', top: '100%', left: '50%', transform: 'translate(-50%, 130%)' }}
        >
          {user.substring(0, user.indexOf('@'))}
        </Typography>
      ) : (
        <SvgIcon inheritViewBox sx={StyledLogo} aria-label="logo">
          <Logo className="logo" />
        </SvgIcon>
      )}
    </AppBar>
  );
};

export default Header;
