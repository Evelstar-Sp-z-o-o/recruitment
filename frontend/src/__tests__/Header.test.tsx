import React from 'react';
import { useLocation } from 'react-router-dom';

import { vi } from 'vitest';

import Header from '@/src/components/Organisms/Header/Header';
import { setUser, store } from '@/src/store';
import { getInitials } from '@/src/utils/helpers/getInitials';
import { renderWithProviders } from '@/src/utils/providers/renderWithProviders';
import '@testing-library/jest-dom';
import { screen, fireEvent } from '@testing-library/react';

vi.mock('react-router-dom', async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    useLocation: vi.fn(),
  };
});

const mockToggleMenu = vi.fn();

describe('Header', () => {
  const renderComponent = (pathname = '/') => {
    (useLocation as vi.Mock).mockReturnValue({ pathname });

    return renderWithProviders(<Header toggleMenu={mockToggleMenu} />);
  };

  it('renders logo if not on profile page', () => {
    renderComponent();

    expect(screen.getByLabelText(/logo/i)).toBeInTheDocument();
  });

  it('does not render logo if on profile page', () => {
    store.dispatch(setUser('test@example.com'));
    renderComponent('/profile');

    expect(screen.queryByLabelText(/logo/i)).not.toBeInTheDocument();
  });

  it('renders user avatar if user is logged in', () => {
    store.dispatch(setUser('test@example.com'));
    renderComponent();

    expect(screen.getByAltText('User').getAttribute('src')).toMatch(/userIcon.svg/);
  });

  it('renders unknown icon if user is not logged in', () => {
    store.dispatch(setUser(null));
    renderComponent();

    expect(screen.getByAltText('User').getAttribute('src')).toMatch(/noUserIcon.svg/);
  });

  it('displays user initials on profile page', () => {
    store.dispatch(setUser('test@example.com'));
    const user = 'test@example.com';
    renderComponent('/profile');

    expect(screen.getByText(getInitials(user))).toBeInTheDocument();
    expect(screen.getByText('test')).toBeInTheDocument();
  });

  it('calls toggleMenu when avatar button is clicked', () => {
    store.dispatch(setUser(null));
    renderComponent();

    fireEvent.click(screen.getByRole('button'));

    expect(mockToggleMenu).toHaveBeenCalledWith(true);
  });

  it('should match snapshot', () => {
    const snapshot = renderWithProviders(<Header toggleMenu={mockToggleMenu} />);
    expect(snapshot).toMatchSnapshot();
  });
});
