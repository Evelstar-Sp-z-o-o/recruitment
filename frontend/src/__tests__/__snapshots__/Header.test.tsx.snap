// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorDefault MuiAppBar-positionStatic css-owe5nn-MuiPaper-root-MuiAppBar-root"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-x6g243-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular css-1qjb60v-MuiAvatar-root"
          >
            <img
              alt="User"
              class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
              src="/src/assets/noUserIcon.svg"
            />
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <svg
          aria-hidden="true"
          aria-label="logo"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1o780gf-MuiSvgIcon-root"
          focusable="false"
        >
          <svg
            class="logo"
            fill="none"
            viewBox="0 0 325 325"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M107.5 92.5H53L107.5 38M107.5 92.5H150.5H162M107.5 92.5L53 38H107.5M107.5 92.5L162 38M107.5 38L162 92.5M107.5 38H162M162 92.5L216.5 38M162 92.5H216.5M216.5 38L271 92.5H216.5M216.5 38H162M216.5 38H271L216.5 92.5M162 38L216.5 92.5"
              stroke="black"
              stroke-width="16"
            />
            <path
              d="M162 100.5C162 100.5 162 195.5 162 251C162 287.031 117 285.5 117 285.5"
              stroke="black"
              stroke-width="45"
            />
          </svg>
        </svg>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorDefault MuiAppBar-positionStatic css-owe5nn-MuiPaper-root-MuiAppBar-root"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-x6g243-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular css-1qjb60v-MuiAvatar-root"
        >
          <img
            alt="User"
            class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="/src/assets/noUserIcon.svg"
          />
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <svg
        aria-hidden="true"
        aria-label="logo"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1o780gf-MuiSvgIcon-root"
        focusable="false"
      >
        <svg
          class="logo"
          fill="none"
          viewBox="0 0 325 325"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M107.5 92.5H53L107.5 38M107.5 92.5H150.5H162M107.5 92.5L53 38H107.5M107.5 92.5L162 38M107.5 38L162 92.5M107.5 38H162M162 92.5L216.5 38M162 92.5H216.5M216.5 38L271 92.5H216.5M216.5 38H162M216.5 38H271L216.5 92.5M162 38L216.5 92.5"
            stroke="black"
            stroke-width="16"
          />
          <path
            d="M162 100.5C162 100.5 162 195.5 162 251C162 287.031 117 285.5 117 285.5"
            stroke="black"
            stroke-width="45"
          />
        </svg>
      </svg>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
